File: Middle-Of-Linked-List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

let middleNode = function (head) {
  let targetLen = Math.ceil(getLength(head) / 2);
  return returnAtPosition(head, targetLen);
};

let getLength = function (node) {
  let length = 0;
  while (node.next) {
    length += 1;
    node = node.next;
  }
  return length;
};

let returnAtPosition = function (node, targetLen) {
  for (let i = 0; i < targetLen; i++) {
    node = node.next;
  }
  return node;
};

File: middle_of_linked_list.py
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        count = 0
        tmp = head
        while tmp:
            count+=1
            tmp = tmp.next
            
        middle = count//2
        l = 0
        while l < middle:
            head = head.next
            l+=1
        return head
